### PR TITLE
Created TOGGLE changer

### PR DESCRIPTION
**Pull request by DeltaRays**
## Description
Created TOGGLE changer, updated all expressions to work with it, and added TOGGLE to EffChange

**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

File: src/main/java/ch/njol/skript/effects/EffChange.java [hunk 1]
```diff
@@ -1,18 +1,18 @@
 /**
- *   This file is part of Skript.
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/effects/EffChange.java:2]: This PR has **A LOT** of unnecessary line changes. 
A PR should only include changes which are relevant to the idea behind the PR itself.
Fixing spaces in javadocs/copyrights, and also fixes other spacing is not relevant to this PR itself.

File: src/main/java/ch/njol/skript/effects/EffChange.java [hunk 2]
```diff
@@ -56,43 +56,44 @@
...
-		"remove every iron tool from the player",
-		"remove all minecarts from {entitylist::*}",
-		"# delete:",
-		"delete the block below the player",
-		"clear drops",
-		"delete {variable}",
-		"# reset:",
-		"reset walk speed of player",
-		"reset chunk at the targeted block"})
-@Since("1.0 (set, add, remove, delete), 2.0 (remove all)")
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/effects/EffChange.java:79]: Here is another spot.
While I do agree these spaces are annoying and do need to be updated (at some point) they cause a lot of change to this PR and in turn create a lot more for us to review. 
While this clearly is simple to see what was changed, in some cases code could potentially be snuck in under our noses if we were to brush over this quickly.

Comment (DeltaRays) [reply src/main/java/ch/njol/skript/effects/EffChange.java:79]: Yes, but first of all I wouldn't sneak code in and secondly, you'd have to do the check were I to make another pull request about the line changes anyways

Comment (DeltaRays) [reply src/main/java/ch/njol/skript/effects/EffChange.java:79]: Moreover, I don't have access to a version of the files without line changes so I'd need to remake everything from scratch

Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/effects/EffChange.java:79]: > Yes, but first of all I wouldn't sneak code in 

This is not an issue with you, this is an issue in general. And as a general coding practice as well as security, this should be reverted.

> Moreover, I don't have access to a version of the files without line changes so I'd need to remake everything from scratch

1) This shouldn't have been changed in the first place
2) The original version can be copy/pasted from the repo.

Comment (DeltaRays) [reply src/main/java/ch/njol/skript/effects/EffChange.java:79]: Yes, the original version can be but I don't want to have to spend hours remaking everything 😕

**Comment by APickledWalrus** [comment]
While the actual implementation is fine, I think the fixes could be more thought out (see changes like this in #3241)
- Making sure expressions use a switch statement over an 'against' if check where applicable in `acceptChange`
- Possibly making sure `change` has a `default` section that asserts false.

File: src/main/java/ch/njol/skript/expressions/ExprAI.java [hunk 3]
```diff
@@ -50,12 +50,14 @@ public Boolean convert(LivingEntity entity) {
...
 	public Class<?>[] acceptChange(Changer.ChangeMode mode) {
-		return mode == Changer.ChangeMode.SET ? CollectionUtils.array(Boolean.class) : null;
+		return (mode == Changer.ChangeMode.SET  || mode == Changer.ChangeMode.TOGGLE) ? CollectionUtils.array(Boolean.class) : null;
 	}
 	
 	@Override
 	public void change(Event event, @Nullable Object[] delta, Changer.ChangeMode mode) {
 		if (delta == null || delta[0] == null) {
+			if(mode == Changer.ChangeMode.TOGGLE)
+				for (LivingEntity entity : getExpr().getArray(event)) entity.setAI(!entity.hasAI());
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprAI.java:60]: The method for setting AI should be on the next line.

File: src/main/java/ch/njol/skript/expressions/ExprBiome.java [hunk 4]
```diff
@@ -74,6 +74,7 @@ public Biome convert(final Location l) {
 	@Override
 	@Nullable
 	public Class<?>[] acceptChange(final ChangeMode mode) {
+		if(mode == ChangeMode.TOGGLE) return null;
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprBiome.java:77]: the return should be on the next line.
As well there should be a space between if and the parenthesis

File: src/main/java/ch/njol/skript/expressions/ExprCreeperMaxFuseTicks.java [hunk 5]
```diff
@@ -52,7 +52,7 @@ public Number convert(LivingEntity e) {
 	@Override
 	@Nullable
 	public Class<?>[] acceptChange(final ChangeMode mode) {
-		if (mode == ChangeMode.REMOVE_ALL)
+		if (mode == ChangeMode.REMOVE_ALL ||mode == ChangeMode.TOGGLE)
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprCreeperMaxFuseTicks.java:55]: there should be a space between | and mode

File: src/main/java/ch/njol/skript/expressions/ExprFlightMode.java [hunk 6]
```diff
@@ -48,17 +48,21 @@ public Boolean convert(final Player player) {
...
 		return null;
 	}
 
 	@Override
 	public void change(Event event, @Nullable Object[] delta, Changer.ChangeMode mode) {
-		boolean state = mode != Changer.ChangeMode.RESET && delta != null && (boolean) delta[0];
-		for (Player player : getExpr().getArray(event)) {
-			player.setAllowFlight(state);
+		if(mode == Changer.ChangeMode.TOGGLE){
+			for (Player player : getExpr().getArray(event)) player.setAllowFlight(!player.getAllowFlight());
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprFlightMode.java:60]: the method for setting allow flight should be on the next line

File: src/main/java/ch/njol/skript/expressions/ExprFlightMode.java [hunk 7]
```diff
@@ -48,17 +48,21 @@ public Boolean convert(final Player player) {
...
 		}
 		return null;
 	}
 
 	@Override
 	public void change(Event event, @Nullable Object[] delta, Changer.ChangeMode mode) {
-		boolean state = mode != Changer.ChangeMode.RESET && delta != null && (boolean) delta[0];
-		for (Player player : getExpr().getArray(event)) {
-			player.setAllowFlight(state);
+		if(mode == Changer.ChangeMode.TOGGLE){
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprFlightMode.java:59]: there should be a space between if and ( as well as between ) {

File: src/main/java/ch/njol/skript/expressions/ExprGlidingState.java [hunk 8]
```diff
@@ -60,14 +60,16 @@ protected String getPropertyName() {
...
-		if (mode == ChangeMode.SET || mode == ChangeMode.RESET)
+		if (mode == ChangeMode.SET || mode == ChangeMode.RESET || mode == ChangeMode.TOGGLE)
 			return new Class[] {Boolean.class};
 		return null;
 	}
 	
 	@Override
 	public void change(final Event e, final @Nullable Object[] delta, final ChangeMode mode) throws UnsupportedOperationException {
-		for (final LivingEntity entity : getExpr().getArray(e))
+		if(mode == ChangeMode.TOGGLE)
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprGlidingState.java:70]: There should be a space between if and (

File: src/main/java/ch/njol/skript/expressions/ExprGlidingState.java [hunk 9]
```diff
@@ -60,14 +60,16 @@ protected String getPropertyName() {
...
+			for (final LivingEntity entity : getExpr().getArray(e)) entity.setGliding(!entity.isGliding());
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprGlidingState.java:71]: The method for setting gliding should be on the next line.

File: src/main/java/ch/njol/skript/expressions/ExprGlowing.java [hunk 10]
```diff
@@ -58,14 +58,18 @@ protected String getPropertyName() {
...
-			entity.setGlowing(delta == null ? false : (Boolean) delta[0]);
+		if (mode == ChangeMode.TOGGLE)
+			for (final Entity entity : getExpr().getArray(e))
+				entity.setGlowing(!entity.isGlowing());
+		else
+			for (final Entity entity : getExpr().getArray(e))
+				entity.setGlowing(delta == null ? false : (Boolean) delta[0]);
 	}
-}
+}
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprGlowing.java:75]: final line should not have been removed

File: src/main/java/ch/njol/skript/expressions/ExprGravity.java [hunk 11]
```diff
@@ -60,14 +60,18 @@ protected String getPropertyName() {
...
-			entity.setGravity(delta == null ? true : (Boolean) delta[0]);
+		if (mode == ChangeMode.TOGGLE)
+			for (final Entity entity : getExpr().getArray(e))
+				entity.setGravity(!entity.hasGravity());
+		else
+			for (final Entity entity : getExpr().getArray(e))
+				entity.setGravity(delta == null ? true : (Boolean) delta[0]);
 	}
-}
+}
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprGravity.java:77]: final line should not have been removed

File: src/main/java/ch/njol/skript/expressions/ExprPassenger.java [hunk 12]
```diff
@@ -114,6 +114,7 @@ public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelaye
 			return new Class[] {Entity[].class, EntityData[].class}; // To support more than one entity
 		if (mode == ChangeMode.SET)
 			return new Class[] {Entity.class, EntityData.class};
+		if (mode == ChangeMode.TOGGLE) return null;
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprPassenger.java:117]: return should be on the next line.

File: src/main/java/ch/njol/skript/expressions/ExprShooter.java [hunk 13]
```diff
@@ -76,6 +76,7 @@ public LivingEntity convert(final Projectile p) {
 	public Class<?>[] acceptChange(final ChangeMode mode) {
 		if (mode == ChangeMode.SET)
 			return new Class[] {LivingEntity.class};
+		if (mode == ChangeMode.TOGGLE) return null;
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprShooter.java:79]: return should be on the next line.

File: src/main/java/ch/njol/skript/expressions/ExprTarget.java [hunk 14]
```diff
@@ -110,6 +110,7 @@ public boolean setTime(final int time) {
 	public Class<?>[] acceptChange(final ChangeMode mode) {
 		if (mode == ChangeMode.SET || mode == ChangeMode.DELETE)
 			return CollectionUtils.array(LivingEntity.class);
+		if (mode == ChangeMode.TOGGLE) return null;
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprTarget.java:113]: return should be on the next line.

File: src/main/java/ch/njol/skript/expressions/ExprVehicle.java [hunk 15]
```diff
@@ -103,6 +103,7 @@ protected String getPropertyName() {
 		if (mode == ChangeMode.SET) {
 			return new Class[] {Entity.class, EntityData.class};
 		}
+		if (mode == ChangeMode.TOGGLE) return null;
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprVehicle.java:106]: return should be on the next line.

**Comment by ShaneBeee** [comment]
My other concern with this PR is I'm not sure how much you have tested.
There is a LOT of changes to a lot of expressions. That said I do feel there should be a regression test made to include tests for each expression that has been changed. 
This way we know whether or not the things changed actually do work.

File: src/main/java/ch/njol/skript/effects/EffChange.java (hunk 2 above)
Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/effects/EffChange.java:79]: This still needs to be resolved. I do not want to hear the excuse "I don't want to have to spend hours remaking everything"
1) You shouldn't change things in the first place that aren't part of the PR
2) You just have to highlight this section and hit <tab>

Comment (DeltaRays) [reply src/main/java/ch/njol/skript/effects/EffChange.java:79]: I literally copied that section from the skript master branch (https://raw.githubusercontent.com/SkriptLang/Skript/master/src/main/java/ch/njol/skript/effects/EffChange.java) and pasted it in EffChange and it didn't change anything so I thought it wasn't an issue. Moreover, that was not an excuse but it was the truth.

**Comment by DeltaRays** [comment]
Regarding the last thing, I don't see why it's needed for expressions that have just been updated to not allow TOGGLE, but you're right, I'll do it for the ones which now use the toggle changer, too.
One thing, how would I check for something related to players?

**Comment by DeltaRays** [comment]
Also, how would I check src/main/java/ch/njol/skript/expressions/ExprEventCancelled.java

File: src/main/java/ch/njol/skript/expressions/ExprSaturation.java [hunk 16]
```diff
@@ -73,7 +73,7 @@ public Number get(final Player player) {
 	@Nullable
 	@Override
 	public Class<?>[] acceptChange(Changer.ChangeMode mode) {
-		return (mode != ChangeMode.REMOVE_ALL) ? CollectionUtils.array(Number.class) : null;
+		return (mode != ChangeMode.REMOVE_ALL && mode != ChangeMode.TOGGLE) ? CollectionUtils.array(Number.class) : null;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprSaturation.java:76]: You should use a switch statement here (like in other spots)

File: src/main/java/ch/njol/skript/expressions/ExprGravity.java [hunk 17]
```diff
@@ -60,14 +60,18 @@ protected String getPropertyName() {
...
+		if (mode == ChangeMode.SET || mode == ChangeMode.RESET || mode == ChangeMode.TOGGLE)
 			return new Class[] {Boolean.class};
 		return null;
 	}
 	
 	@Override
 	public void change(final Event e, final @Nullable Object[] delta, final ChangeMode mode) throws UnsupportedOperationException {
-		for (final Entity entity : getExpr().getArray(e))
-			entity.setGravity(delta == null ? true : (Boolean) delta[0]);
+		if (mode == ChangeMode.TOGGLE)
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprGravity.java:70]: I think you should put curly brackets on the if / else statements here. *Maybe* the for loops too.

File: src/main/java/ch/njol/skript/effects/EffChange.java [hunk 18]
```diff
@@ -310,9 +319,11 @@ public String toString(final @Nullable Event e, final boolean debug) {
...
 				return "reset " + changed.toString(e, debug);
+			case TOGGLE:
+				return "toggle " + changed.toString(e, debug);
 		}
 		assert false;
 		return "";
 	}
 	
-}
+}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/effects/EffChange.java:329]: New line? :D

File: src/main/java/ch/njol/skript/expressions/ExprAI.java [hunk 19]
```diff
@@ -50,12 +50,21 @@ public Boolean convert(LivingEntity entity) {
...
+				return CollectionUtils.array(Boolean.class);
+			default:
+				return null;
+		}
 	}
 	
 	@Override
 	public void change(Event event, @Nullable Object[] delta, Changer.ChangeMode mode) {
 		if (delta == null || delta[0] == null) {
+			if(mode == Changer.ChangeMode.TOGGLE)
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprAI.java:65]: There should be a space after the `if` statement. There should also be curly brackets around the `if` statement IMO

File: src/main/java/ch/njol/skript/expressions/ExprEnchantingExpCost.java [hunk 20]
```diff
@@ -72,7 +72,7 @@ public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelaye
 	@Override
 	@Nullable
 	public Class<?>[] acceptChange(ChangeMode mode) {
-		if (mode == ChangeMode.RESET || mode == ChangeMode.DELETE || mode == ChangeMode.REMOVE_ALL)
+		if (mode == ChangeMode.RESET || mode == ChangeMode.DELETE || mode == ChangeMode.REMOVE_ALL || mode == ChangeMode.TOGGLE)
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprEnchantingExpCost.java:75]: Switch statement here

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprSaturation.java (hunk 16 above)
Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprSaturation.java:76]: oh no
as I said the TOGGLE changer is the only one I didn't look into 😬

Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/expressions/ExprSaturation.java:76]: Oh, right haha. Might as well leave these here anyway >:)

**Comment by DeltaRays** [comment]
I've chosen to close this PR for the moment.

**Pull request closed without merging**